File: solutions/35.search-insert-position/src/binary_search.ts
/*
 * @lc app=leetcode.cn id=35 lang=typescript
 *
 * https://leetcode.cn/problems/search-insert-position/
 */

// @lc code=start
function searchInsert(nums: number[], target: number): number {
    let left = 0;
    let right = nums.length - 1;
    while (left <= right) {
        const mid = (left + right) >> 1;
        const x = nums[mid];
        if (x < target) {
            left = mid + 1;
        } else if (x > target) {
            right = mid - 1;
        } else {
            return mid;
        }
    }
    return left;
}
// @lc code=end

export { searchInsert };

File: solutions/35.search-insert-position/src/binary_search.rs
/*
 * @lc app=leetcode.cn id=35 lang=rust
 *
 * https://leetcode.cn/problems/search-insert-position/
 */

pub struct Solution;

// @lc code=start
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let mut left = 0;
        let mut right = nums.len() as i32 - 1;
        while left <= right {
            let mid = (left + right) / 2;
            let x = nums[mid as usize];
            if x < target {
                left = mid + 1;
            } else if x > target {
                right = mid - 1;
            } else {
                return mid;
            }
        }
        left
    }
}
// @lc code=end
